lib/wasix/src/syscalls/wasix/closure_prepare.rs

```diff
@@ -0,0 +1,315 @@
+use std::path::PathBuf;
+
+use super::*;
+use crate::{state::WasmLoader, syscalls::*};
+use wasmer::{imports, wat2wasm, FunctionType, Table, Type};
+
+// TODO: Move to wasix-types
+#[repr(u8)]
+enum WasmValueType {
+    I32 = 0,
+    I64 = 1,
+    F32 = 2,
+    F64 = 3,
+}
+impl WasmValueType {
+    fn new(value: u8) -> Result<Self, Errno> {
+        match value {
+            0 => Ok(Self::I32),
+            1 => Ok(Self::I64),
+            2 => Ok(Self::F32),
+            3 => Ok(Self::F64),
+            _ => Err(Errno::Inval),
+        }
+    }
+    fn as_u8(self) -> u8 {
+        self as u8
+    }
+    fn size(&self) -> u64 {
+        match self {
+            Self::I32 => 4,
+            Self::I64 => 8,
+            Self::F32 => 4,
+            Self::F64 => 8,
+        }
+    }
+    fn name(&self) -> &str {
+        match self {
+            Self::I32 => "i32",
+            Self::I64 => "i64",
+            Self::F32 => "f32",
+            Self::F64 => "f64",
+        }
+    }
+}
+
+
+
+/// TODO: write proper documentation for this function
+
+#[instrument(level = "trace", skip_all, fields(path = field::Empty), ret)]
+pub fn closure_prepare<M: MemorySize>(
+    mut ctx: FunctionEnvMut<'_, WasiEnv>,
+    // The backing function that will receive the serialized parameters
+    // It needs to have the following signature:
+    // * pointer to the values
+    // * pointer to the results
+    // * pointer to the userdata
+    backing_function_id: u32,
+    // The ID of the function that will be registered
+    function_id: u32,
```

Comment:
Why the hardcoded function ID?

Reply:
The `closure_*` syscalls work by first allocating a index in the function table and then preparing a closure at that index. The libffi interface requires us to do this in two steps.